src/main/java/com/elytradev/concrete/inventory/gui/widget/WDirectionalPanel.java

```diff
@@ -0,0 +1,194 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package com.elytradev.concrete.inventory.gui.widget;
+
+/**
+ * Like a JPanel with a BoxLayout
+ */
+public class WDirectionalPanel extends WPanel {
+	public static final int DEFAULT_SPACING = 4;
+	
+	private final Direction direction;
+	private int spacing = DEFAULT_SPACING;
+	
+	public WDirectionalPanel(Direction direction) {
+		this.direction = direction;
+	}
+	
+	/**
+	 * Sets the spacing between widgets of this panel.
+	 * <p>
+	 * This method changes layout-related information, and therefore,
+	 * invalidates the panel hierarchy. If the panel has already been
+	 * validated, the hierarchy must be re-validated thereafter in order to
+	 * reflect the changes.
+	 *
+	 * @see #invalidate
+	 * @see #validate
+	 */
+	public void setSpacing(int spacing) {
+		this.spacing = spacing;
+		invalidate();
+	}
+	
+	@Override
+	public void pack() {
+		int directionSpace = 0;
+		int centeredSpace = 0;
+		for (int i = 0; i < getWidgetCount(); i++) {
+			WWidget w = getWidget(i);
+			if (i > 0) {
+				directionSpace += spacing;
+			}
+			directionSpace += direction.getDirectionSpace(w);
+			centeredSpace = Math.max(direction.getCenteredSpace(w), centeredSpace);
+		}
+		direction.setSize(this, directionSpace, centeredSpace);
+	}
+	
+	@Override
+	protected void layout() {
+		int unresizable = 0;
+		int numResizable = 0;
+		for (int i = 0; i < getWidgetCount(); i++) {
+			WWidget w = getWidget(i);
+			if (w.canResize()) {
+				numResizable++;
+			} else {
+				unresizable += direction.getDirectionSpace(w);
+			}
+		}
+		int resizeSpace = direction.getDirectionSpace(this) - unresizable;
+		int resizeEach;
+		if (numResizable > 0) {
+			resizeEach = resizeSpace / numResizable;
+			if (resizeEach < 18) {
+				resizeEach = 18; //Don't squish things beyond a minimum reasonable size! Better for them to overlap :/
+			}
+		} else {
+			resizeEach = 0; //None of them are resizable, so set this to an arbitrary value
+		}
+		
+		int centerline = direction.getCenteredSpace(this) / 2;
+		int curOffset = 0;
+		for (int i = 0; i < getWidgetCount(); i++) {
+			WWidget w = getWidget(i);
+			if (w.canResize()) {
+				direction.setSize(w, resizeEach, direction.getCenteredSpace(this));
+			}
+			direction.setLocation(w, curOffset, centerline - (direction.getCenteredSpace(w) / 2));
+			curOffset += direction.getDirectionSpace(w) + spacing;
+		}
+		
+		//Layout children after parents
+		super.layout();
+	}
+	
+	public static enum Direction {
+		/**
+		 * <ul>
+		 * <li>If all children are non-resizable, they will wind up equally
+		 * spaced horizontally, with their vertical position centered on the
+		 * panel's centerline.
```

Comment:
You should close tags in javadocs.
IIRC Java 8's new javadoc complainer tool will moan about this.